### ISSUE TITLE
SPIFF inconsistencies

### ISSUE THREAD

**Issue by malaterre** [team-charls/charls#185]
This is a general comment about the CharLS C-API with regard to SPIFF header inconsistencies. Attaching a simple (dumb) example:

![inconsistant_spiff](https://user-images.githubusercontent.com/228803/161555596-5cd38fef-45b1-4c0e-907c-bf75849302d3.jpg)

IMHO the easy fix should be around here:

```
    jpegls_decoder& read_header(CHARLS_OUT std::error_code& ec) noexcept
    {
        ec = charls_jpegls_decoder_read_header(decoder_.get());
        if (ec == jpegls_errc::success)
        {
            ec = charls_jpegls_decoder_get_frame_info(decoder_.get(), &frame_info_);
            if (ec == jpegls_errc::success && decoder.spiff_header_has_value_)
              {
              ec = charls_jpegls_is_spiff_consistant_with_frame_info(decoder_.spiff_header_, &frame_info_);
              }
        }
        return *this;
    }
```

Not clear if this is the right fix though. The behavior of `charls_jpegls_decoder_read_spiff_header` would need also to be clarified (eg. what if `compression_type != charls::spiff_compression_type::jpeg_ls`).

**Comment by malaterre** [comment]
```
static bool charls_jpegls_is_spiff_consistant_with_frame_info(const charls_spiff_header* spiff_header,
                                                              const charls_frame_info* frame_info)
{
    return spiff_header->component_count == frame_info->component_count && spiff_header->height == frame_info->height &&
           spiff_header->width == frame_info->width && spiff_header->bits_per_sample == frame_info->bits_per_sample;
}
```

**Comment by vbaderks** [comment]
At the moment the SPIFF header is not validated. As CharLS only reads this SPIFF header and doesn't use its values this is the minimal implementation. The advantage is that an invalid SPIFF header does not prevent successful decoding an image.
I think there are 2 directions to improve it. 
1. Document that the SPIFF header is not validated. It is then up to the application code to verify the SPIFF header.
2. Add a method that validates the SPIFF header and check if it is consistent with the frame_info.

Option 2 would make it easier for applications to rely on the library to validate the SPIFF header, which can be considered untrusted data.

The only drawback I see is that the spiff_header.bits_per_sample field is limited to only the values: 1,2,4, 8, 12 and 16. (Table F.1 in ISO/IEC 10918-3) and the JPEG-LS standards allows the range [2,16]. Existing 10 and 14 bit images would become invalid.
Excluding this check can however be documented.

**Comment by malaterre** [comment]
I can see an interest for a "fix spiff marker" application. So I like your option (2) above.

**Comment by vbaderks** [comment]
Method has been added with PR #189

**Comment by malaterre** [comment]
@vbaderks Thanks a lot for the fix !

For some reason I do no see the following comment in the documentation for `charls_validate_spiff_header`:

> The only drawback I see is that the spiff_header.bits_per_sample field is limited to only the values: 1,2,4, 8, 12 and 16. (Table F.1 in ISO/IEC 10918-3) and the JPEG-LS standards allows the range [2,16]. Existing 10 and 14 bit images would become invalid. Excluding this check can however be documented.

Is the above still accurate ?

**Comment by vbaderks** [comment]
Thanks for the feedback. Documentation was missing, resolved with f1d2f691b1e8a52f4acff7f8b2dbe957a8abde6e
